Implement timeShift natively

shanson7 opened the pull request:
**Describe your changes**
Implement timeShift natively

**Testing performed**
test files added
A/B testing against graphite using https://github.com/shanson7/mtcmptest

**Additional context**
There are a few scenarios that are a little strange:

1. `resetToEnd` does not seem to work when data is retrieved from MT. I think it's due to padding the data to the requested end.
2. `alignToDST` is hard to test on UTC servers :) It uses the server timezone (which I think is misleading to users) and that may be different in the Mt deploy than the graphite deploy

Dieterbe committed:
run chaos-cluster test in circleCI

this seems useful especially for PR's that affect clustering

Dieterbe committed:
add warnings about hard to find docker-compose issues

Dieterbe committed:
don't use -V: circleCI still doesn't support it

Dieterbe committed:
make sure chaos tests don't get stuck pulling images in CircleCI

replay committed:
rename metaTagIndex to metaTagHierarchy

replay committed:
refactor index locks

replay committed:
deduplicate meta tag records when returning from meta tag hierarchy

replay committed:
refactor methods idsByTagQuery and query methods

replay committed:
make benchmark report allocs

replay committed:
use accessor methods everywhere

Dieterbe committed:
use Archives type for cleaner way of planning requests

* simpler more useful type with optimized fields for planning requests
* no need to make special exceptions everywhere to get the metric
  RawInterval when archive=0

Dieterbe committed:
clarification

shanson7 committed:
support separate from/to windows

Dieterbe committed:
typo: GroupData also used for singles

shanson7 committed:
Remove unused variable

shanson7 committed:
Add test for independent planning

shanson7 committed:
Add timeShift func and basic tests

shanson7 committed:
timeShift stable

shanson7 committed:
Allocate array up front if needed

fqiu7 commented on expr/validator.go:
```diff
@@ -37,6 +37,17 @@ func IsIntervalString(e *expr) error {
 
+func IsSignedIntervalString(e *expr) error {
+	durStr := e.str
+	if durStr[0] == '-' {
```
why not use
```
if (durStr[0] == '-' || durStr[0] == '+') {
    durStr = durStr[1:]
}
```

shanson7 replied:
Good catch, this was copy/pasta from other code that had more handling.

fqiu7 commented on expr/func_timeshift.go:
```diff
@@ -0,0 +1,153 @@
+
+func addOffset(orig uint32, offset int) uint32 {
+	if offset < 0 {
+		orig -= uint32(offset * -1)
```
so what happens if this result is negative (orig will be an incorrect positive value here)?

shanson7 replied:
Yes see comment at https://github.com/bloomberg/metrictank/pull/89/files/01ea8dd74d455203529e2456290e579b6cb12dc0#diff-a08ba41f2effbe7b76d8c3a9f891cd75R46

fqiu7 commented on expr/func_timeshift.go:
```diff
@@ -0,0 +1,153 @@
+		return nil, err
+	}
+
+	negativeOffset := -s.shiftOffset
```
is s.shiftOffset a signed offset?

shanson7 replied:
yes, it is an int

fqiu7 commented on expr/pool.go:
```diff
@@ -12,3 +16,13 @@ var pointSlicePool *sync.Pool
+	out := pointSlicePool.Get().([]schema.Point)
+	if cap(out) < minSize {
+		// TODO - return `out` to pool? Could just fetch it again next iteration
+		out = make([]schema.Point, 0, minSize)
```
so in this case, out is not fetched out of the pool, it is more like calling make() locally

shanson7 replied:
yes

shanson7 committed:
Fix test cases

shanson7 committed:
More concise code in validator

agao48 commented on expr/func_timeshift_test.go:
```diff
@@ -0,0 +1,226 @@
+			if (bothNaN || p.Val == exp.Datapoints[j].Val) && p.Ts == exp.Datapoints[j].Ts {
+				continue
+			}
+			t.Fatalf("case %q: output point %d - expected %v got %v", name, j, exp.Datapoints[j], p)
```
no way to use `equalOutput`?

shanson7 committed:
use equalOutput in test

fqiu7 reviewed (approved):
LGTM

agao48 committed:
Don't modify series in place

agao48 committed:
Skeleton minMax function

agao48 committed:
Implement min_max

agao48 committed:
Add minmax tests

shanson7 committed:
Add tests - zero series input, does not modify input

robert-milan committed:
re-add recording language to community call

agao48 committed:
minMax -> stable

robert-milan committed:
Merge pull request #1790 from grafana/update-call-md

re-add recording language to community call

agao48 committed:
Flesh out invert test
* split out input unchanged check
* 0 series, 2 series

agao48 committed:
Extra tests for minmax function

agao48 committed:
make tests more consistent with new timeshift

agao48 committed:
make tests more consistent with new timeshift tests

replay committed:
implement /metrics/expand endpoint

shanson7 committed:
Dedup resolve series requests

replay committed:
use errgroup

Dieterbe committed:
Rework Graphite docs

* double down our focus on Graphite, put that more front and center.
* document removal of constantLine().
* document render formats difference. fix #1468
* update some of the items

Dieterbe committed:
remove irrelevant items

replay committed:
document /metrics/expand

shanson7 committed:
Update MEMBERS.md

shanson7 committed:
Update GOVERNANCE.md

Dieterbe committed:
Apply suggestions from code review

Co-Authored-By: Robert Milan <42070645+robert-milan@users.noreply.github.com>

Dieterbe committed:
fix field names

Dieterbe committed:
Merge pull request #1796 from grafana/rework-graphite-docs

Rework graphite docs

Dieterbe committed:
Merge pull request #1798 from grafana/fix-metadata-field-names

fix field names in render metadata: consolidate -> consolidator

shanson7 committed:
Store Req as copy not pointer

Dieterbe committed:
Merge pull request #1797 from grafana/shanson-onboarding

Shanson onboarding

shanson7 committed:
add clarifying comment

Dieterbe committed:
Merge pull request #1794 from bloomberg/dedup_series

Dedup resolve series requests

shanson7 committed:
Add overlap checker to DataMap for tests

shanson7 committed:
Mock adds series to datamap

shanson7 committed:
Additional test helpers

shanson7 committed:
Round 1 of func test normalization and improvement

shanson7 committed:
Round 2 of test normalization

Dieterbe committed:
Merge pull request #1738 from grafana/use-archives-for-cleaner-archinterval-exceptions

Use new archive type for cleaner non-standard archinterval handling

replay committed:
better naming

replay committed:
fix race in test

replay committed:
fix sorting expressions by cost

with and without meta tag support

shanson7 committed:
Round 3 of test normalization

shanson7 committed:
Round 4 of test normalization

shanson7 committed:
Add tag comparison check

shanson7 committed:
Fix functions that modify series in place

shanson7 committed:
gofmt

shanson7 committed:
Fix tests

Dieterbe committed:
cleanup tracker code + support printing when each matcher matches

Dieterbe committed:
assert that zookeer/kafka has come up. might be needed for circleCI

shanson7 committed:
Use pointers so plan.Clean() works

Dieterbe committed:
alignment: better, more comprehensive testing

Dieterbe committed:
simpler align.{Backward,Forward}IfNotAligned: no need to branch

Dieterbe committed:
Merge pull request #1804 from bloomberg/fix_plan_pool

Use pointers so plan.Clean() works

Dieterbe committed:
in circleCI, need to be able to connect to 'kafka' from the host

Dieterbe committed:
use less intensive workload so that circleCI can still handle it

replay committed:
improve FAQ

Dieterbe committed:
Merge pull request #1801 from grafana/chaos-cluster-in-circleci-2

run chaos-cluster test in circleci (continuation of #1431)

robert-milan committed:
Merge pull request #1803 from grafana/better-alignment-testing

Faster, simpler alignment methods, better testing

Dieterbe committed:
clarify chaos test better

Dieterbe committed:
fixes. use more relatable example using default retention

Dieterbe committed:
Merge pull request #1808 from grafana/improve_faq

Update FAQ to explain why query engine may choose low resolution aggregate

Dieterbe committed:
document storage-schemas.conf better

this is an import of a change that was already improved/merged on our
website

Dieterbe committed:
update crng package versions

Dieterbe committed:
refer to http-api docs + use same url syntax in all examples

Dieterbe committed:
Merge pull request #1810 from grafana/docs-cloud-describe-storage-schemas-conf-better

Docs cloud describe storage schemas conf better

shanson7 committed:
Add test for prepending points

shanson7 committed:
Add more NaN to pad series

Dieterbe committed:
Merge pull request #1813 from bloomberg/fix_normalizeto

Dieterbe committed:
Merge pull request #1809 from grafana/clarify-chaos-test

clarify chaos test better

Dieterbe committed:
move metric builder to more proper package

Dieterbe committed:
clearer

Dieterbe committed:
support customizing the printing of the "number" of the metric

Dieterbe committed:
allow to choose the builder instead of hardcoding an arbitrary one

Dieterbe committed:
support new "daily-sine" value policy

shanson7 committed:
Use consReq in merge key

fkaleo committed:
If the client disconnects while processing a query that needs to be forwarded to graphite, return http code 499 instead of default 502 by implementing a custom proxy error handler

replay committed:
add missing RUnlock()

Dieterbe committed:
Merge pull request #1792 from bloomberg/min_max

Implement minMax function natively

Dieterbe committed:
Merge pull request #1791 from bloomberg/invert

Implement invert function natively

Dieterbe committed:
Merge pull request #1815 from grafana/mt-fakemetrics-various

Mt-fakemetrics: configurable builders + "daily-sine" value policy

replay committed:
fix locking around id by tag query lookup

shanson7 committed:
Cleaner diff for equalTags

Dieterbe committed:
Merge pull request #1766 from grafana/fix_1765

Fix Meta tag inverted filtering not working

replay committed:
Merge pull request #1550 from grafana/20191125_index_refactor_for_more_granular_locks

index refactor for more granular locks

shanson7 committed:
Fix consolidation bug

shanson7 committed:
Fix up test for removeEmptySeries

shanson7 committed:
Make xFilesFactor optional, make graphite compatible

Dieterbe committed:
Merge pull request #1823 from bloomberg/fix_removeEmptySeries

Fix removeEmptySeries

fkaleo committed:
Merge pull request #1821 from grafana/499_when_client_disconnects

Return 499 http code instead of 502 when client disconnect during a render query with graphite proxying

Dieterbe committed:
fix & improve comments

Dieterbe committed:
fix datafeed loop starting to lag behind if ticks are missed

Dieterbe committed:
fix metrics for different orgs affecting each other's timestamps

BEFORE: per-org logic affects ts for subsequent orgs
./build/mt-fakemetrics feed --orgs 10 --mpo 1 --stdout                                                                                                              ⏎
params: metricName=some.id.of.a.metric.%d, orgs=10, mpo=1, period=1, flush=1000, offset=0, speedup=1, stopAtNow=false
per org:         each 1s, flushing 10 metrics so rate of 10 Hz. (10 total unique series)
times   10 orgs: each 1s, flushing 10 metrics so rate of 10 Hz. (10 total unique series)
2020/05/15 11:44:54 [I] starting listener on :6764
1 some.id.of.a.metric.1 (tags: []) 0.604660 1589532294
2 some.id.of.a.metric.1 (tags: []) 0.664560 1589532295
3 some.id.of.a.metric.1 (tags: []) 0.000000 1589532296
4 some.id.of.a.metric.1 (tags: []) 0.525096 1589532297
5 some.id.of.a.metric.1 (tags: []) 0.872726 1589532298
6 some.id.of.a.metric.1 (tags: []) 3.606488 1589532299
7 some.id.of.a.metric.1 (tags: []) 1.285583 1589532300
8 some.id.of.a.metric.1 (tags: []) 0.954175 1589532301
9 some.id.of.a.metric.1 (tags: []) 0.283034 1589532302
10 some.id.of.a.metric.1 (tags: []) 0.679085 1589532303
1 some.id.of.a.metric.1 (tags: []) 1.219121 1589532304
2 some.id.of.a.metric.1 (tags: []) 3.994713 1589532305
3 some.id.of.a.metric.1 (tags: []) 2.638028 1589532306
4 some.id.of.a.metric.1 (tags: []) 0.000000 1589532307
5 some.id.of.a.metric.1 (tags: []) 0.865335 1589532308
6 some.id.of.a.metric.1 (tags: []) 1.571461 1589532309
7 some.id.of.a.metric.1 (tags: []) 1.424954 1589532310
8 some.id.of.a.metric.1 (tags: []) 8.777175 1589532311
9 some.id.of.a.metric.1 (tags: []) 4.163660 1589532312
10 some.id.of.a.metric.1 (tags: []) 0.000000 1589532313
1 some.id.of.a.metric.1 (tags: []) 1.039597 1589532314
2 some.id.of.a.metric.1 (tags: []) 0.544156 1589532315
3 some.id.of.a.metric.1 (tags: []) 2.962065 1589532316
4 some.id.of.a.metric.1 (tags: []) 0.507081 1589532317
5 some.id.of.a.metric.1 (tags: []) 5.520234 1589532318
6 some.id.of.a.metric.1 (tags: []) 0.880543 1589532319
7 some.id.of.a.metric.1 (tags: []) 1.788723 1589532320
8 some.id.of.a.metric.1 (tags: []) 8.792252 1589532321
9 some.id.of.a.metric.1 (tags: []) 1.111447 1589532322
10 some.id.of.a.metric.1 (tags: []) 0.966060 1589532323
1 some.id.of.a.metric.1 (tags: []) 0.932846 1589532324
2 some.id.of.a.metric.1 (tags: []) 0.000000 1589532325
3 some.id.of.a.metric.1 (tags: []) 1.280474 1589532326
4 some.id.of.a.metric.1 (tags: []) 8.072928 1589532327
5 some.id.of.a.metric.1 (tags: []) 0.000000 1589532328
6 some.id.of.a.metric.1 (tags: []) 0.272512 1589532329
7 some.id.of.a.metric.1 (tags: []) 1.853974 1589532330
8 some.id.of.a.metric.1 (tags: []) 2.435678 1589532331
9 some.id.of.a.metric.1 (tags: []) 0.710907 1589532332
10 some.id.of.a.metric.1 (tags: []) 5.195916 1589532333

AFTER: all orgs have their own state, bug is fixed
./build/mt-fakemetrics feed --orgs 10 --mpo 1 --stdout
params: metricName=some.id.of.a.metric.%d, orgs=10, mpo=1, period=1, flush=1000, offset=0, speedup=1, stopAtNow=false
per org:         each 1s, flushing 10 metrics so rate of 10 Hz. (10 total unique series)
times   10 orgs: each 1s, flushing 10 metrics so rate of 10 Hz. (10 total unique series)
2020/05/15 12:19:58 [I] starting listener on :6764
1 some.id.of.a.metric.1 (tags: []) 0.604660 1589534398
2 some.id.of.a.metric.1 (tags: []) 0.664560 1589534398
3 some.id.of.a.metric.1 (tags: []) 0.000000 1589534398
4 some.id.of.a.metric.1 (tags: []) 0.525096 1589534398
5 some.id.of.a.metric.1 (tags: []) 0.872726 1589534398
6 some.id.of.a.metric.1 (tags: []) 3.606488 1589534398
7 some.id.of.a.metric.1 (tags: []) 1.285583 1589534398
8 some.id.of.a.metric.1 (tags: []) 0.954175 1589534398
9 some.id.of.a.metric.1 (tags: []) 0.283034 1589534398
10 some.id.of.a.metric.1 (tags: []) 0.679085 1589534398
1 some.id.of.a.metric.1 (tags: []) 1.219121 1589534399
2 some.id.of.a.metric.1 (tags: []) 3.994713 1589534399
3 some.id.of.a.metric.1 (tags: []) 2.638028 1589534399
4 some.id.of.a.metric.1 (tags: []) 0.000000 1589534399
5 some.id.of.a.metric.1 (tags: []) 0.865335 1589534399
6 some.id.of.a.metric.1 (tags: []) 1.571461 1589534399
7 some.id.of.a.metric.1 (tags: []) 1.424954 1589534399
8 some.id.of.a.metric.1 (tags: []) 8.777175 1589534399
9 some.id.of.a.metric.1 (tags: []) 4.163660 1589534399
10 some.id.of.a.metric.1 (tags: []) 0.000000 1589534399
1 some.id.of.a.metric.1 (tags: []) 1.039597 1589534400
2 some.id.of.a.metric.1 (tags: []) 0.544156 1589534400
3 some.id.of.a.metric.1 (tags: []) 2.962065 1589534400
4 some.id.of.a.metric.1 (tags: []) 0.507081 1589534400
5 some.id.of.a.metric.1 (tags: []) 5.520234 1589534400
6 some.id.of.a.metric.1 (tags: []) 0.880543 1589534400
7 some.id.of.a.metric.1 (tags: []) 1.788723 1589534400
8 some.id.of.a.metric.1 (tags: []) 8.792252 1589534400
9 some.id.of.a.metric.1 (tags: []) 1.111447 1589534400
10 some.id.of.a.metric.1 (tags: []) 0.966060 1589534400
1 some.id.of.a.metric.1 (tags: []) 0.932846 1589534401
2 some.id.of.a.metric.1 (tags: []) 0.000000 1589534401

Dieterbe committed:
correctly stop publishing datafeed when we reach 'now'

fix #1824

Dieterbe committed:
use time of invocation as "stop at now" value, not wallclock

this is typically what people want, but we can reintroduce the old
behavior later if somebody misses it.

Dieterbe committed:
Merge pull request #1825 from grafana/cleanup-fakemetrics

mt-fakemetrics cleanup and fixes

shanson7 committed:
Normalize nonNegativeDerivative

shanson7 committed:
Normalize minmax tests

shanson7 committed:
Fix accidental consolidator reset in minmax

shanson7 committed:
Update wording for CoW

shanson7 committed:
Rename getTargetSeries and getQuerySeries

shanson7 committed:
Add comment about overlap detection

shanson7 committed:
Clean up divideSeries tests a bit

shanson7 committed:
Sorting needs CoW semantic too

shanson7 committed:
init dataMap outside of the mock calls

robert-milan committed:
add utf8 validation checks to incoming metrics

robert-milan committed:
update zoom link for community call

Dieterbe committed:
Merge pull request #1832 from grafana/update-community-docs

update zoom link for community call

Dieterbe committed:
Merge pull request #1800 from bloomberg/expr_test_normalization

Fix processing funcs changes affecting other targets + refactor all func tests to test for clobbering

replay committed:
add benchmark for adding & deleting meta records

replay committed:
synchronously lookup ids in enricher add/del meta record

replay committed:
add comments

replay committed:
use callback to submit to enricher before release read lock

replay committed:
minor fix: keep holding lock

robert-milan committed:
use less allocations in checks

robert-milan committed:
update metrics.md for invalid utf8

replay committed:
remove outdated comments

Dieterbe committed:
set points-return more accurately

by looking at the actual response
instead of the known-inaccurate estimate from before the plan run

Dieterbe committed:
Merge pull request #1835 from grafana/more-accurate-points-return

set points-return more accurately

robert-milan committed:
update tests and add more comments

shanson7 committed:
Use non-blocking write to flushed

shanson7 committed:
ran gofmt

replay committed:
Merge pull request #1834 from grafana/fix_deadlock_1833

Fix deadlock 1833

robert-milan committed:
consolidate validation types into one

robert-milan committed:
update invalid_tag metric

robert-milan committed:
Merge pull request #1831 from grafana/reject-invalid-utf8

add utf8 validation checks to incoming metrics

robert-milan committed:
update changelog to include breaking change from #1831

Dieterbe committed:
Merge pull request #1841 from grafana/upadte-changelog-for-1831

update changelog to include breaking change from #1831

Dieterbe committed:
document why calling consolidation.GetAggFunc() in these cases is safe

Dieterbe committed:
consolidation.Consolidate needs a valid consolidator. make explicit

Dieterbe committed:
models.Req.Consolidator is never 0, mark it as such

Dieterbe committed:
expr.NormalizeTo assure valid series.Consolidator just like Plan.Run()

Dieterbe committed:
Merge pull request #1843 from grafana/consolidator-none

assure that consolidation.Consolidate always gets a non-None consolidator

robert-milan committed:
Merge pull request #1819 from bloomberg/fix_1807

Merge should consider consReq

robert-milan committed:
Merge pull request #1837 from bloomberg/fix_1836

Use non-blocking write to flushed

Dieterbe committed:
add script to format stackdump

DanCech committed:
cloud docs are now in the website repo

Dieterbe committed:
Merge pull request #1844 from grafana/scripts-dev-format-stackdump-sh

add script to format stackdump

Dieterbe committed:
document consolidation in response path + refine normalization naming

disambiguate between on-demand runtime normalization and after-fetch
runtime normalization

Dieterbe committed:
remove docs GH action

DanCech committed:
Merge pull request #1846 from grafana/remove-cloud-docs

cloud docs are now in the website repo

Dieterbe committed:
canonical and pre-canonical clarifications

Dieterbe committed:
improve newRequestContext() docs

Dieterbe committed:
improve render request handling docs

replay committed:
make metric parsing respect spaces in tags

replay committed:
add unit tests to check for issue 1672

Dieterbe committed:
introduce unit tests to see whether series are canonical...

...at least for a couple different functions, and also for summarize in
the !alignToFrom case.

Dieterbe committed:
summarize(): in !alignToFrom case, return output in canonical format

See https://github.com/grafana/metrictank/issues/1811 for more details
Before, it was common for output to contain an extraneous point that
lies before 'from' (and wouldn't be rendered by Grafana in graph
panels), which made the output invalid.

This would result in trouble when

* combining such a series with another series
  (https://github.com/grafana/metrictank/issues/1811)
* or when needing normalization
  (https://github.com/grafana/metrictank/issues/1845)

This effectively slightly changes the output format, but is more robust.

Dieterbe committed:
refactor: isolate out "make pre-canonical" logic

Dieterbe committed:
Merge pull request #1849 from grafana/fix-1811-1845

Make summarize() output canonical in the !alignToFrom case

replay committed:
better comment

Co-authored-by: Dieter Plaetinck <dieter@grafana.com>

replay committed:
Merge pull request #1848 from grafana/fix_spaces

Fix functions aliasByNode and aliasSub when a tag contains a space char

RichiH committed:
Adopt governance and workflow structure from Grafana

Signed-off-by: Richard Hartmann <richih@richih.org>

Dieterbe committed:
improve TestParse

* add test case for deeply nested query
* also show leftover data
* some formatting fixes

Dieterbe committed:
support piped syntax

Dieterbe committed:
more accurately handle incomplete query call statements

Dieterbe committed:
deduplicate plan.Reqs

it's trivial to demonstrate that before this change, a query like
target=foo&target=foo results in duplicates in plan.Reqs, which
leads to duplicates in ReqMap during executePlan().

Dieterbe committed:
document render path series (de)duplication and impact on series reuse

Dieterbe committed:
ReqMap.Dump() output fix

Dieterbe committed:
document series reuse stuff better

Dieterbe committed:
utility function for the pool to get appropriately sized slice

Dieterbe committed:
fix: Identical series causes double consolidation

apply COW when altering points slice during runtime normalization
fix #1807

Dieterbe committed:
typo fixes

Co-authored-by: Sean Hanson <shanson7@bloomberg.net>

Dieterbe committed:
only attempt to get from pool once

any further attempts would likely yield the same result

Dieterbe committed:
Merge pull request #1855 from grafana/fixes-for-duplicate-series

duplicate series: deduplicate + fix runtime normalization

Dieterbe committed:
update changelog

Dieterbe committed:
add random ticker

Dieterbe committed:
add a "Containers" mode to fakemetrics with configurable churn

Dieterbe committed:
Merge pull request #1859 from grafana/mt-fakemetrics-containers-churn

add a "Containers" mode to fakemetrics with configurable churn

RichiH committed:
Add Code of Conduct

Signed-off-by: Richard Hartmann <richih@richih.org>

dependabot[bot] committed:
Bump json from 1.8.3 to 2.3.1 in /scripts

Bumps [json](https://github.com/flori/json) from 1.8.3 to 2.3.1.
- [Release notes](https://github.com/flori/json/releases)
- [Changelog](https://github.com/flori/json/blob/master/CHANGES.md)
- [Commits](https://github.com/flori/json/compare/v1.8.3...v2.3.1)

Signed-off-by: dependabot[bot] <support@github.com>

Dieterbe committed:
Merge pull request #1862 from grafana/dependabot/bundler/scripts/json-2.3.1

Bump json from 1.8.3 to 2.3.1 in /scripts

robert-milan committed:
add aggregateWithWildcards and tests

robert-milan committed:
fix signature and constructors

robert-milan committed:
fix unit tests

Dieterbe committed:
clarify that pcre is not supported

RichiH committed:
Update maintainers

Signed-off-by: Richard Hartmann <richih@richih.org>

Dieterbe committed:
Merge pull request #1853 from grafana/richih/governance

Adopt governance and workflow structure from Grafana

Dieterbe committed:
Merge pull request #1857 from grafana/update-changelog

update changelog

Dieterbe committed:
changelog tweaks

Dieterbe committed:
Merge pull request #1865 from grafana/changelog-tweaks

changelog tweaks

Dieterbe committed:
Merge pull request #1854 from grafana/pipe-syntax

Render query pipe syntax

Dieterbe committed:
update changelog

Dieterbe committed:
fix "sgauge" to "gauge". this was probably a typo

Dieterbe committed:
add metric to track number of goroutines. fix #1198

Dieterbe committed:
update dashboard for Grafana Grafana v7.1.1 (3039f9c3bd)

Dieterbe committed:
add panel showing number of goroutines

Dieterbe committed:
update a config that was never updated

Dieterbe committed:
Merge pull request #1867 from grafana/config-tweak

update a config file that was never updated

Dieterbe committed:
clarify & tweak comments

Dieterbe committed:
fix reporting of archives

Dieterbe committed:
clarifications

Dieterbe committed:
clean up code

Dieterbe committed:
cleanup: ratio is not a percentage

Dieterbe committed:
leverage in-flight requests of other replicas to possibly overcome shard query errors

fix #1817

Dieterbe committed:
Merge pull request #1869 from grafana/fix-spec-exec-err-but-inflight

leverage in-flight requests of other replicas to possibly overcome shard query errors

Dieterbe committed:
simpler

Dieterbe committed:
add aggregateSeriesWithWildcards()

Dieterbe committed:
simplify + do it like Graphite does

Dieterbe committed:
more thorough input copy. see #1872

Dieterbe committed:
Merge pull request #1866 from grafana/stats-tweaks

monitoring of number of goroutines and other stats tweaks

Dieterbe committed:
Merge pull request #1868 from grafana/fix-archive-stats

fix reporting of archives

shanson7 committed:
Archive metrics separately for singles

Dieterbe committed:
Merge pull request #1739 from bloomberg/more_time

Support multiple targets that act on different time windows

shanson7 committed:
Fix function name

shanson7 committed:
Add timeshift to docs

Dieterbe committed:
Merge pull request #1863 from grafana/add-sum-series-with-wildcards

Adds aggregateWithWildcards (and sum / multiply / avg)

Dieterbe committed:
exclude customer-impacting tickets from stalebot

Dieterbe committed:
add logo. fix #1488

Dieterbe committed:
clarify pointSlicePool / DataMap expecations/behavior

robert-milan committed:
Merge pull request #1877 from grafana/add-logo

add logo. fix #1488

Dieterbe committed:
make logo smaller

Dieterbe committed:
attempt 2

Dieterbe committed:
tweak

Dieterbe committed:
Merge pull request #1878 from grafana/logo-resize

make logo smaller

Dieterbe committed:
allow 2 different kinds of series lifecycle management for normalize functions

Dieterbe committed:
Update api/dataprocessor.go

Co-authored-by: Sean Hanson <shanson7@bloomberg.net>

Dieterbe committed:
Merge pull request #1879 from grafana/fix-double-put

Fix same models.Series getting added to the sync.Pool (or datamap) twice

replay committed:
Merge pull request #1795 from grafana/metrics_expand

implement /metrics/expand endpoint

Dieterbe committed:
need to use "export for external use" for Grafana.com upload

Dieterbe committed:
fix datasource for 2 panels

Dieterbe committed:
fix GOGC and Goroutines display for multi instances

Dieterbe committed:
Merge pull request #1883 from grafana/dashboard-export-with-datasource-fix

Dashboard export with datasource fix

Dieterbe committed:
add note about consolidateBy

Dieterbe committed:
tweak

robert-milan committed:
update confid.md

Dieterbe committed:
Merge pull request #1876 from grafana/stale-exclude-customer-impacting

exclude customer-impacting tickets from stalebot

Dieterbe committed:
Merge pull request #1864 from grafana/clarify-pcre-not-supported

clarify that pcre is not supported

Dieterbe committed:
Merge branch 'master' into graphite-document-consolidateBy

Dieterbe committed:
Merge pull request #1884 from grafana/graphite-document-consolidateBy

add note about consolidateBy

shanson7 committed:
Merge branch 'master' into timeShift_take2

shanson7 committed:
clean up spaces

shanson7 committed:
Clean up spaces

shanson7 commented:
Merged in https://github.com/grafana/metrictank/pull/1873
